ecosystem/sep-0030.md

```diff
@@ -0,0 +1,479 @@
+- Alternatively, the client uses [SEP-10] to get a [JWT] proving high threshold control of another account that is listed as an alternative identity of the account to be accessed.
+- The client, authenticated as the phone, email, or another account, requests the server to sign a transaction.
+- The server verifies that the [JWT] is valid according to the authentication providers specification.
+- The server verifies that the phone, email, or account in the [JWT] is an alternative identity of the account.
+- The server verifies that the source account of the transaction is the account.
+- The server verifies that the source account of all operations on the transaction are not set or set to the account.
+- The server signs the transaction with the signing secret key.
+- The server responds to the client with the signature.
+
+The client repeats the signing flow with each server it needs to meet the
+accounts threshold to be able to successfully submit a transaction.
+
+The client can use this process to sign a transaction with the [Set Options
+operation] to add a new signer to the account to regain control of an account
+of which they have lost the key, or to gain control of an account shared with
+them.
+
+A client can also use the endpoints in the specification to:
+- Discover registered accounts that it has access to.
+- Update the identities associated with an account.
+- Delete accounts.
+
+## Specification
+
+### Authentication
+
+All endpoints require authentication in the form of a [JWT] token that is passed
+in the `Authorization` header. The [JWT] can be issued by two types of issuers.
+Some endpoints require a specific provider while others should function with
+both. The two issuers are:
+
+- `SEP-10`: A [SEP-10] Web Auth server for proving high threshold control of an
+account. The [SEP-10] server should be configured to only issue [JWT]s if signers
+prove they meet the high threshold on the account.
+- `External`: Defined by the implementer. Proves control of a phone number,
+email address, or other identity.
+
+### Content Type
+
+All endpoints accept in requests the following `Content-Type`s:
+- `application/x-www-form-urlencoded`
+- `application/json`
+
+All endpoints respond with content type:
+- `application/json`
+
+### Errors
+
+If an error occurs when calling any endpoint, an appropriate HTTP status code
+will be returned along with an error response.
+
+#### Status Code
+
+Common HTTP status codes may be returned for a server. In particular the following are expected:
+
+Status Code | Name | Reason
+-----|------|------------
+`400` | Bad Request | The request is invalid in anyway.
+`401` | Unauthorized | No `Authorization` header has been provided or the contents of the header are not accepted as valid.
+`404` | Not Found | The path is not recognized by the server, or the address does not have an account, or the authenticated user does not have permission to access the account.
```

Comment:
Exactly as you say the choice of 404 is because the resource doesn't exist to somebody not authenticated or authenticated and doesn't have access.

Different APIs have interpreted this differently. AWS S3 uses 403 if you try and access anything you don't have access to, even if it actually doesn't exist. GitHub uses 404. I think 404 is more consistent with the actual descriptions of the status codes in our use case. It's also easier to implement because when implementing it as 403 it is really easy to accidentally return 404 in some cases and then leak insight.

It definitely makes the API a little more obtuse as it could be easy to think it doesn't exist if you aren't authenticated properly.